qiskit/transpiler/instruction_durations.py

```diff
@@ -0,0 +1,141 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Durations of instructions, one of transpiler configurations."""
+import warnings
+from typing import Optional, List, Tuple, Union, Iterable
+
+from qiskit.circuit import Barrier, Delay
+from qiskit.circuit import Instruction, Qubit
+from qiskit.providers import BaseBackend
+from qiskit.transpiler.exceptions import TranspilerError
+
+InstructionDurationsType = List[Tuple[str, Optional[Iterable[int]], int]]
+"""List of tuples representing (instruction name, qubits indices, duration)."""
+
+
+class InstructionDurations:
+    """Helper class to provide integer durations for safe scheduling."""
+
+    def __init__(self,
+                 instruction_durations: Optional[InstructionDurationsType] = None,
+                 schedule_dt=None):
+        self.duration_by_name = {}
+        self.duration_by_name_qubits = {}
+        self.schedule_dt = schedule_dt
+        if instruction_durations:
+            self.update(instruction_durations, schedule_dt)
+
+    @classmethod
+    def from_backend(cls, backend: BaseBackend):
+        """Construct the instruction durations from the backend."""
+        if backend is None:
+            return InstructionDurations()
+        # TODO: backend.properties() should tell us all about instruction durations
+        if not backend.configuration().open_pulse:
+            raise TranspilerError("No backend.configuration().dt in the backend")
+
+        dt = backend.configuration().dt  # pylint: disable=invalid-name
+        instruction_durations = []
+        # backend.properties._gates -> instruction_durations
+        for gate, insts in backend.properties()._gates.items():
+            for qubits, props in insts.items():
+                if 'gate_length' in props:
+                    gate_length = props['gate_length'][0]  # Throw away datetime at index 1
+                    duration = round(gate_length / dt)
+                    rounding_error = abs(gate_length - duration * dt)
+                    if rounding_error > 1e-15:
+                        warnings.warn("Duration of %s is rounded to %d dt = %e s from %e"
+                                      % (gate, duration, duration * dt, gate_length),
+                                      UserWarning)
+                    instruction_durations.append((gate, qubits, duration))
+        # To know duration of measures, to be removed
+        inst_map = backend.defaults().instruction_schedule_map
+        all_qubits = tuple(range(backend.configuration().num_qubits))
+        meas_duration = inst_map.get('measure', all_qubits).duration
+        for q in all_qubits:
+            instruction_durations.append(('measure', [q], meas_duration))
+        return InstructionDurations(instruction_durations, dt)
```

Comment:
Move this into `__init__` so we can do `InstructionDurations(backend)`. This could be called `_process_backend(self, backend: BaseBackend)`. If a `backend` is provided, (you could check the input type) then call this on the backend, but rather than returning a new `InstructionDurations`, you update this one.

If you compare the old `CmdDef` to `InstructionScheduleMap`, you'll see that we decided this format is preferable after many conversations.

Reply:
OK, I'll follow the convention of pulse module.

Reply:
I have looked through `InstructionScheduleMap` carefully. I understand `cmd_def: List[Command]` and `InstructionScheduleMap` correspond with `instruction_durations` (list of tuples) and `InstructionDurations`, respectively. That means 
1. InstructionDurations's constructor should not take any argument, i.e. it should be always created as an empty InstructionDurations
2. backend (backend_properties) should return `instruction_durations`
3. InstructionDurations should be constructed from `instruction_durations`
Current interface does not cover (2) yet since we decided not to update backend model in this PR.
I cannot see any significant difference between current `InstructionDurations.from_backend()` and `InstructionDurations(backend)`. Either would be deprecated when `backend_properties` is updated in the future if we force users to create `InstructionDurations` from `instruction_durations`.
I think current interface sufficiently respects what we learnt in pulse module.
Could you elaborate on your concern on current interface?

Reply:
I guess it's fine. Just seems weird to have a "construct from object" method, when the object could just be constructed from that object initially. Seems to be the point of an `__init__` method. I don't care too strongly since the user interface won't be acceptable until this: https://github.com/Qiskit/qiskit-terra/issues/4703